### PR TITLE
Fix #224: make sure unique html id

### LINKED ISSUE #224: Make sure id is unique for each HTML in markdown transform

### PR DESCRIPTION
**Pull request by qinezh**
@vwxyzh @vicancy @chenkennt  @superyyrrzz

### CODE REVIEW

File: src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockRules/MarkdownHeadingBlockRule.cs [hunk 1]
```diff
@@ -24,7 +24,7 @@ public virtual IMarkdownToken TryMatch(IMarkdownParser parser, ref string source
                     t.Rule,
                     t.Context,
                     p.TokenizeInline(match.Groups[2].Value),
-                    Regex.Replace(match.Groups[2].Value.ToLower(), @"[^\w]+", "-"),
+                    Regex.Replace(match.Groups[2].Value.ToLower(), @"[^\p{L}\p{Nd}]+", "-"),
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockRules/MarkdownHeadingBlockRule.cs]: too restricted? {Nd} -> {N} ?

File: src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs [hunk 2]
```diff
@@ -36,5 +38,17 @@ public MarkdownHeadingBlockToken Rewrite(IMarkdownRewriteEngine rewriterEngine)
...
             return new MarkdownHeadingBlockToken(Rule, Context, c, Id, Depth, RawMarkdown);
         }
+
+        public MarkdownHeadingBlockToken RewriteId(Dictionary<string, int> idTable)
+        {
+            if (idTable.ContainsKey(Id))
+            {
+                var newId = string.Concat(Id, "-",idTable[Id]);
+                idTable[Id]++;
+                return new MarkdownHeadingBlockToken(Rule, Context, Content, newId, Depth, RawMarkdown);
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs]: Will the newID collide with an existing ID?

File: src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs [hunk 3]
```diff
@@ -36,5 +38,17 @@ public MarkdownHeadingBlockToken Rewrite(IMarkdownRewriteEngine rewriterEngine)
             }
             return new MarkdownHeadingBlockToken(Rule, Context, c, Id, Depth, RawMarkdown);
         }
+
+        public MarkdownHeadingBlockToken RewriteId(Dictionary<string, int> idTable)
+        {
+            if (idTable.ContainsKey(Id))
+            {
+                var newId = string.Concat(Id, "-",idTable[Id]);
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs]: add space after `"-",`

**Comment by superyyrrzz** [comment]
:+1:

**Comment by qinezh** [comment]
Ping..

File: src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs [hunk 4]
```diff
@@ -36,5 +38,31 @@ public MarkdownHeadingBlockToken Rewrite(IMarkdownRewriteEngine rewriterEngine)
             }
             return new MarkdownHeadingBlockToken(Rule, Context, c, Id, Depth, RawMarkdown);
         }
+
+        public MarkdownHeadingBlockToken RewriteId(Dictionary<string, int> idTable)
+        {
+            var newId = GenerateNewId(idTable, Id);
+            if (string.Equals(newId, Id))
+            {
+                return null;
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs]: what does return null mean?

File: src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs [hunk 5]
```diff
@@ -36,5 +38,31 @@ public MarkdownHeadingBlockToken Rewrite(IMarkdownRewriteEngine rewriterEngine)
...
+            }
+            return new MarkdownHeadingBlockToken(Rule, Context, Content, newId, Depth, RawMarkdown);
+        }
+
+        private string GenerateNewId(Dictionary<string, int> idTable, string Id)
+        {
+            if (idTable.ContainsKey(Id))
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs]: use TryGetValue

File: src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs (hunk 4 above)
Comment (qinezh) [inline src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs]: `return null` means this token will be ignored in this part of rewrite process

File: src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs [hunk 6]
```diff
@@ -36,5 +38,32 @@ public MarkdownHeadingBlockToken Rewrite(IMarkdownRewriteEngine rewriterEngine)
...
+                var newId = string.Concat(Id, "-", count);
+                idTable[Id] = count + 1;
+                return GenerateNewId(idTable, newId);
+            }
+            else
+            {
+                idTable[Id] = 0;
+                return Id;
+            }
+
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.MarkdownLite/Basic/BlockTokens/MarkdownHeadingBlockToken.cs]: remove this line

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**